### DISCUSSION TITLE
Framing error, unexpected byte: 63

### DISCUSSION THREAD

**Discussion by cozyvision** [php-amqplib/php-amqplib#1037] [Q&A]
**Version(s) affected**: 3.2.0

**Description**  
"Framing error, unexpected byte: 63"
```
[hex]:
0000  41 4D 51 50 00 00 09 01                            AMQP....

waiting for 10,10
waiting for a new frame
closing input
closing socket
Framing error, unexpected byte: 63
#0 /var/www/html/rabbitmq/app/Vendor/PhpAmqpLib/Connection/AbstractConnection.php(658): PhpAmqpLib\Connection\AbstractConnection->wait_frame()
#1 /var/www/html/rabbitmq/app/Vendor/PhpAmqpLib/Channel/AbstractChannel.php(234): PhpAmqpLib\Connection\AbstractConnection->wait_channel()
#2 /var/www/html/rabbitmq/app/Vendor/PhpAmqpLib/Channel/AbstractChannel.php(351): PhpAmqpLib\Channel\AbstractChannel->next_frame()
#3 /var/www/html/rabbitmq/app/Vendor/PhpAmqpLib/Connection/AbstractConnection.php(266): PhpAmqpLib\Channel\AbstractChannel->wait()
#4 /var/www/html/rabbitmq/app/Vendor/PhpAmqpLib/Connection/AbstractConnection.php(237): PhpAmqpLib\Connection\AbstractConnection->connect()
#5 /var/www/html/rabbitmq/app/Vendor/PhpAmqpLib/Connection/AMQPStreamConnection.php(64): PhpAmqpLib\Connection\AbstractConnection->__construct()
#6 /var/www/html/rabbitmq/app/Console/Command/Task/ProcessSmtpTask.php(56): 
```

**Comment by ramunasd** [comment]
Very likely You are connecting to wrong port/service.

**Comment by cozyvision** [comment]
i tried to enter some other(wrong) IP and port combination.

Wrong IP
`stream_socket_client(): Unable to connect to tcp://10.8.0.1:5672 (Connection timed out)`
Wrong Port
`stream_socket_client(): Unable to connect to tcp://127.0.0.1:56727 (Connection refused)`

But with the right IP and port(i am using the same IP,port, username and password in my java code and its working from the same server).

**Comment by ramunasd** [comment]
Any errors in rabbitmq logs?

**Comment by cozyvision** [comment]
RabbitMQ logs "client unexpectedly closed TCP connection"

```
2022-10-25 11:57:13.586152+00:00 [info] <0.28585.0> accepting AMQP connection <0.28585.0> (127.0.0.1:44438 -> 127.0.0.1:5672)
2022-10-25 11:57:13.588942+00:00 [warning] <0.28585.0> closing AMQP connection <0.28585.0> (127.0.0.1:44438 -> 127.0.0.1:5672):
2022-10-25 11:57:13.588942+00:00 [warning] <0.28585.0> client unexpectedly closed TCP connection
```

**Comment by lukebakken** [comment]
@cozyvision Your original message is not enough information for us to assist you. Please see [this document](https://github.com/rabbitmq/support-tools/blob/main/docs/Reporting_RabbitMQ_Issues.md) for the information we need to assist.

Please do the following:
* Upgrade to 3.4.0 - https://github.com/php-amqplib/php-amqplib/releases/tag/v3.4.0
* Tell us what RabbitMQ and Erlang version you are using
* Is your PHP application connecting **DIRECTLY** to RabbitMQ or is there a firewall, proxy, load balancer or other network device in-between?
* Has this issue _always_ happened or did it start happening recently?
* Does this issue happen _every_ time your application makes a request to RabbitMQ or is it sporadic?

**Comment by cozyvision** [comment]
I already upgraded to 3.4.0(just before raising the issue), but the issue is still same.

RabbitMQ 3.11.2
Erlang 25.1.1

We connect directly to RabbitMQ as both application and RabbitMQ are on same server(we are using 127.0.0.1/localhost)

The application was working fine since a few months, but yesterday i restarted my service, and after that its no longer working

The issue happens everytime i run it.

However i can still use the same application when i run it on another server(keeping rabbitmq on same machine as its now)

Here is the details in requested format https://projects.cozyvision.com/rabbitmq-env-mq-01.cozyvision.com-20221025-153947.tgz

**Comment by lukebakken** [comment]
@cozyvision thanks for the information. There's nothing really out of the ordinary except that your application can't connect on one server. It appears to be happening during the connect/AMQP handshake phase.

Some more questions and observations:
* "yesterday i restarted my service, and after that its no longer working" - you mean you restarted your PHP application, correct? Did _anything_ else change other than restarting it?
* How is your application written? Is it a console app, web app, etc?
* You mention that you have a Java application running on the same server that does not have any issues. Is that still correct?
* I do see `|-firewalld,371241 -s /usr/sbin/firewalld --nofork --nopid` in the process tree output. Have you configured this at all? I can't imagine that it would affect localhost connections though.
* The `dmesg -t` output shows that the OOM killer has been active on this system, which indicates that it may not have enough memory.

Since you can run the same PHP application on another server my guess at this point is that this library is not the cause of your issue but it is something in your environment.

**Comment by cozyvision** [comment]
- The application was running since 30+ days, and in between i compiled PHP(same version PHP 8.0.20), as we were planning to use this library with PHP Parallel. But we did not restarted our application, since everything was running smoothly even after above process.

- After this issue, we removed the compiled version and reinstalled the pre-compiled version, but still the issue persists.
 
- Our Application is built on Cakephp 2 and we use CakePHP console.
 
- The Java application is still working perfectly and able to receive and process messages. My other servers are using the same PHP library to publish the messages to this server.
 
- Yes, we have firewalld configured(but as you mentioned localhost is not effected by it, moreover the JAVA application is able to connect in same environment, so firewall should not be the issue).
 
- Yes, the OOM was activated, since we were trying a few things with PHP parallel in last few days.
 
- Yes the library is not the issue its likely the environment.

We tried debugging in this library, and found that the code is expecting $ch variable to be 0xCE i.e. 206(frame end), but the value is actually 63(which is invalid).

![image](https://user-images.githubusercontent.com/7001175/197831148-317b7152-3842-41dd-b5b7-25beb4beb798.png)

**Comment by cozyvision** [comment]
Finally i was able to make it work :-)

But do not know exactly what was the issue.

I removed all PHP and related modules.

rebooted the server, 

cleaned the installation

reinstalled Pre-compiled PHP

and it started working.

Thank you for your time and support.
